### PR TITLE
Add eventing e2e tests on s390x architecture

### PR DESCRIPTION
**Pull request by barthy1**
**What this PR does, why we need it**:

This PR proposes the standard way to run Knative tests on non-Intel architectures, as first step with s390x eventing e2e tests. The approach can be applied to other architectures and components. 
The idea that main job is running on Intel architecture, as any other standard test job, but the actual tests will be executed on another k8s arch-specific cluster. From non-Intel architecture k8s cluster and image registry should be provided.
The details:
- the project and test images are built for another architecture (with ko parameter `--platform=linux/<arch>`), despite the actual build is done on x86_64
- images are pushed to special registry (`KO_DOCKER_REPO` parameter)
- the deployment after build is done on remote k8s arch-specific cluster (`KUBECONFIG` parameter points to non-Intel architecture k8s cluster) 
- the tests are running also on remote k8s arch-specific cluster (`KUBECONFIG` parameter parameter points to non-Intel architecture k8s cluster) 

The s390x eventing Prow job:
- is the daily job to run eventing e2e tests on s390x hardware architecture
- it expects from Prow to be able to use `s390x-cluster1` secret to get access to s390x k8s cluster(config file) and image registry(registry.crt) for temporary test images (secrets should be added by Prow maintainer?)

`s390x` hardware is ready and accessible via kubeconfig for k8s cluster. Registry is also up and ready.

The proposed job is tested using mkpj and phaino tools as local Prow job running on Intel hardware with actual tests running on s390x k8s cluster.

config-generator functionality is extended to specify volume/volume mount/env variable for the job (to be able to use secrets with k8s config and registry certificate).

### CODE REVIEW

**Comment by barthy1** [comment]
hmm, I guess I cannot just add directly to `config.yaml`, instead I need to use `config_knative.yaml. Will change my PR with necessary updates.

**Comment by coryrc** [comment]
> it expects from Prow to be able to use s390x-registry and s390x-kubeconfig secrets to get access to s390x k8s cluster and image registry for temporary test images (secrets should be added by Prow maintainer?)

That's easy. How often will they need to be updated? I could see churn on the kubeconfig? I wonder if we could grant you the ability to modify the secret. I think currently we can't but it'd be nice.

> job expects that special service is running on the Prow cluster to hide k8s cluster ip address with hostname(setup should be installed by Prow maintainer?).

I'm pretty sure IPs get printed out in error messages so they wouldn't end up very hidden? But I also don't fully understand this part so I'm probably missing the point.

File: tools/config-generator/main.go
Comment (coryrc) [inline tools/config-generator/main.go:304]: AFAICT you aren't using readwrite volumes, is this in anticipation of something else?

File: tools/config-generator/main.go
Comment (coryrc) [inline tools/config-generator/main.go:306]: Secrets are only readonly so should be an OR?

**Review by coryrc — CHANGES_REQUESTED** [review]

File: tools/config-generator/main.go
Comment (barthy1) [reply tools/config-generator/main.go:304]: the kubeconfig is mounted to `/root/.kube`, and this dir should be readwrite, because extra files are written to it. With readonly setup, the erros occurs
```
error: open /root/.kube/config.lock: read-only file system
```

Comment (coryrc) [reply tools/config-generator/main.go:304]: Oh I see, the absence of readonly has made them mounted readwrite. Makes sense, could you throw a comment in somewhere with this info?

File: tools/config-generator/main.go
Comment (barthy1) [reply tools/config-generator/main.go:306]: well, secrets can be mounted as readwrite, as it is done in current proposal, so actually I guess here should be just check for readonly. Anyway, good point to update, thank you!

Comment (coryrc) [reply tools/config-generator/main.go:306]: Oh, I see now the intention was for this to only be for mount secrets as RW directories... it's confusing to understand in the other places. I think it'd be cleaner if instead of isSecret and isReadOnly variables, we had an enum:

```
Volume
SecretRO
SecretRW
```

or, I see little harm in all secrets being mounted as RW, which is even simpler! It's not like a job could harm another job via that method.

**Comment by barthy1** [comment]
> That's easy. How often will they need to be updated? I could see churn on the kubeconfig? I wonder if we could grant you the ability to modify the secret. I think currently we can't but it'd be nice.

The cluster is stable, so far experience shows that cleanup after tests does its work and nothing needs to be updated/recreated. However of course some unexpected things can happen. So I'd say, yes, the kubeconfig might be updated sometimes, but not frequently. If you can find some easy way for me to update the secret, it would be just great!

> I'm pretty sure IPs get printed out in error messages so they wouldn't end up very hidden? But I also don't fully understand this part so I'm probably missing the point.

Yep, sorry, I had to be more specific. It's not a problem for me to use concrete ip address, however I think it'd be nice to hide it in some special config, just to be able to replace ip address, without touching job itself. However it's not something really mandatory, so up to reviewers to decide.

File: tools/config-generator/main.go
Comment (barthy1) [reply tools/config-generator/main.go:306]: Well, volumes can be mounted readwrite or readonly, and secret is just one possible source for volume. So for me the readonly and secret are just 2 different areas, I'd not mix them.

> I see little harm in all secrets being mounted as RW,

Best practices recommend not to do that, if possible :) However it will make the changes to the current setup smaller. So from my side, I'd still mount all current secrets readonly, as it is done now (except new ones). However of course it's up to maintainers to decide.

File: tools/config-generator/main.go
Comment (barthy1) [reply tools/config-generator/main.go:304]: sure, will add the comment with description

Comment (barthy1) [reply tools/config-generator/main.go:304]: done

**Comment by coryrc** [comment]
> so I'd say, yes, the kubeconfig might be updated sometimes, but not frequently. If you can find some easy way for me to update the secret, it would be just great!

I'll ask around but we're probably stuck with this for now.

> Yep, sorry, I had to be more specific. It's not a problem for me to use concrete ip address, however I think it'd be nice to hide it in some special config, just to be able to replace ip address, without touching job itself.

Ah, because of the peculiarities of this setup, it'll be far easier to update if you have the IP address in a file in this repository and just have it as an ENV var. Then non-googlers can also update it as needed. Or really, the IP address is explicitly tied to the kubeconfig, is it not? so really both should be in the same secret == 50% fewer secrets to maintain!

**Comment by barthy1** [comment]
@coryrc 

> Ah, because of the peculiarities of this setup, it'll be far easier to update if you have the IP address in a file in this repository and just have it as an ENV var. Then non-googlers can also update it as needed. Or really, the IP address is explicitly tied to the kubeconfig, is it not? so really both should be in the same secret == 50% fewer secrets to maintain!

I've rewrote the code to have all required info in 1 secret - k8s config file, registry.crt file and `KO_DOCKER_REPO` value. You are correct, all 3 pieces should be replaced in case if for instance IP address is changed. So better to keep them together + no changes in job config in case if something should be replaced in those values.

File: tools/config-generator/main.go
Comment (barthy1) [reply tools/config-generator/main.go]: I am deeply sorry, you were correct and I was wrong about mount of secrets and always readonly :) It's always readonly.
Only one (small) excuse for me that my test env bypasses this problem. Anyway, code is updated to resolve this challenge. Now I just copy k8s config to another directory, which is RW.

Comment (coryrc) [reply tools/config-generator/main.go]: Haha, no worries! I just know this config generator is already messy so I try to push back a bit on new features to it.

File: tools/config-generator/main.go
Comment (coryrc) [inline tools/config-generator/main.go:346]: nit: s/Informatiion/Information

File: config/prod/prow/config_knative.yaml
Comment (coryrc) [inline config/prod/prow/config_knative.yaml:882]: I think this would be clearer if it were called "external_cluster"? what do you think?
I had to reread the information/comments a few times to figure out why it looked like this, but `external_cluster` would be pretty clear to me.

**Review by coryrc — CHANGES_REQUESTED** [review]
At some point someone from our team will contact you about setting up a GCP project for [this](https://docs.google.com/document/d/1Jo8dyNiJ1fnYh30M_6Gq2XDU9B3bGFmkD2QokyKYwEg/edit?ts=6055480a), so you will have the secret in your own GCP project and can change it at will. (They will also have better instructions soon so don't feel you need to start with it immediately).

But, to start with you can send me the secret and I can apply it to the cluster. My username at my company dot com (google.com) if you want to e-mail or slack it or... I'm open to whatever is easy :).

File: tools/config-generator/main.go
Comment (barthy1) [reply tools/config-generator/main.go]: fixed

File: config/prod/prow/config_knative.yaml
Comment (barthy1) [reply config/prod/prow/config_knative.yaml]: yep, `external_cluster` sounds clearer. Thank you for suggestion!

**Comment by coryrc** [comment]
/lgtm
/approve
/hold

(just holding until the secret is applied -- I think it could be an issue but also maybe not)

**Comment by coryrc** [comment]
Good luck!

**Pull request merged**
